Applications/Spire/Source/Ui/DropDownList.cpp

```diff
@@ -0,0 +1,247 @@
+  m_scroll_area->setWidgetResizable(true);
+  auto main_widget = new QWidget(this);
+  m_layout = new QVBoxLayout(main_widget);
+  m_layout->setContentsMargins({});
+  m_layout->setSpacing(0);
+  m_scroll_area->setWidget(main_widget);
+  set_widget(m_scroll_area);
+  set_items(items);
+  parent->installEventFilter(this);
+}
+
+bool DropDownList::eventFilter(QObject* watched, QEvent* event) {
+  if(watched == parent()) {
+    if(event->type() == QEvent::KeyPress) {
+      auto e = static_cast<QKeyEvent*>(event);
+      if(e->key() == Qt::Key_Up) {
+        if(isVisible()) {
+          focus_previous();
+          return true;
+        }
+      } else if(e->key() == Qt::Key_Down) {
+        if(isVisible()) {
+          focus_next();
+          return true;
+        }
+      } else if(e->key() == Qt::Key_Enter || e->key() == Qt::Key_Return) {
+        if(isVisible() && m_highlight_index) {
+          on_item_selected(get_widget(*m_highlight_index)->get_value(),
+            *m_highlight_index);
+          return true;
+        }
+      }
+    }
+  }
+  return DropDownWindow::eventFilter(watched, event);
+}
+
+void DropDownList::hideEvent(QHideEvent* event) {
+  m_scroll_area->verticalScrollBar()->setValue(0);
+  if(m_highlight_index) {
+    auto item = m_layout->itemAt(*m_highlight_index);
+    if(item != nullptr) {
+      static_cast<DropDownItem*>(item->widget())->reset_highlight();
+    }
+    m_highlight_index = boost::none;
+  }
+}
+
+void DropDownList::keyPressEvent(QKeyEvent* event) {
+  if(event->key() == Qt::Key_Up) {
+    if(isVisible()) {
+      focus_previous();
+    }
+  } else if(event->key() == Qt::Key_Down) {
+    if(isVisible()) {
+      focus_next();
+    }
+  } else if(event->key() == Qt::Key_Escape) {
+    hide();
+  } else if(event->key() == Qt::Key_Enter) {
+    if(m_highlight_index) {
+      on_item_selected(get_widget(*m_highlight_index)->get_value(),
+        *m_highlight_index);
+    }
+  }
+  DropDownWindow::keyPressEvent(event);
+}
+
+connection DropDownList::connect_activated_signal(
+    const ActivatedSignal::slot_type& slot) const {
+  return m_activated_signal.connect(slot);
+}
+
+connection DropDownList::connect_highlighted_signal(
+    const HighlightedSignal::slot_type& slot) const {
+  return m_highlighted_signal.connect(slot);
+}
+
+connection DropDownList::connect_index_selected_signal(
+    const IndexSelectedSignal::slot_type& slot) const {
+  return m_index_selected_signal.connect(slot);
+}
+
+connection DropDownList::connect_value_selected_signal(
+    const ValueSelectedSignal::slot_type& slot) const {
+  return m_value_selected_signal.connect(slot);
+}
+
+QVariant DropDownList::get_value(int index) {
+  if(m_layout->count() - 1 < index) {
+    return QVariant();
+  }
+  return get_widget(index)->get_value();
+}
+
+void DropDownList::insert_item(DropDownItem* item) {
+  m_item_selected_connections.push_back(std::move(
+    item->connect_selected_signal(
+      [=, index = m_layout->count()] (auto value) {
```

Comment:
What if two insertions are happening concurrently? Won't the same item be selected in both cases? If concurrent insertions are not allowed, why copy `index` in the first place?

Reply:
I don't think two insertions could happen concurrently. The issue was that when the lambda is called it was referencing the current m_layout->count(), instead of the index when the lambda was created.